src/mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp

```diff
@@ -0,0 +1,338 @@
+  }
+
+  //! Return the lines sizes.
+  const std::vector<size_t>& LinesSizes() const { return linesSizes; }
+  //! Modify the lines sizes.
+  std::vector<size_t>& LinesSizes() { return linesSizes; }
+
+  //! Return the term frequency type.
+  TfTypes TfType() const { return tfType; }
+  //! Modify the term frequency type.
+  TfTypes& TfType() { return tfType; }
+
+  //! Determine the idf algorithm type (whether it's smooth or not).
+  bool SmoothIdf() const { return smoothIdf; }
+  //! Modify the idf algorithm type (whether it's smooth or not).
+  bool& SmoothIdf() { return smoothIdf; }
+
+  /**
+   * Serialize the class to the given archive.
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */)
+  {
+    ar & BOOST_SERIALIZATION_NVP(tfType);
+    ar & BOOST_SERIALIZATION_NVP(smoothIdf);
+  }
+
+ private:
+  /**
+   * The function calculates the term frequency statistics.
+   *
+   * @tparam ValueType Type of the returned value.
+   *
+   * @param numOccurrences The number of the given token occurrences in
+   *                       the line.
+   * @param numTokens The total number of tokens in the line.
+   */
+  template<typename ValueType>
+  ValueType TermFrequency(const size_t numOccurrences,
+                          const size_t numTokens)
+  {
+    switch (tfType)
+    {
+    case TfTypes::BINARY:
+      return numOccurrences > 0;
+    case TfTypes::RAW_COUNT:
+      return numOccurrences;
+    case TfTypes::TERM_FREQUENCY:
+      return static_cast<ValueType>(numOccurrences) / numTokens;
+    case TfTypes::SUBLINEAR_TF:
+      return std::log(static_cast<ValueType>(numOccurrences)) + 1;
+    default:
+      Log::Fatal << "Incorrect term frequency type!";
+      return 0;
+    }
+  }
+
+  /**
+   * The function calculates the inverse document frequency statistics.
+   *
+   * @tparam ValueType Type of the returned value.
+   *
+   * @param totalNumLines The total number of strings in the input dataset.
+   * @param numOccurrences The number of strings in the input dataset
+   *                       which contain the current token.
+   */
+  template<typename ValueType>
+  ValueType InverseDocumentFrequency(const size_t totalNumLines,
+                                     const size_t numOccurrences)
+  {
+    if (smoothIdf)
+    {
+      return std::log(static_cast<ValueType>(totalNumLines + 1) /
+          (1 + numOccurrences)) + 1.0;
+    }
+    else
+    {
+      return std::log(static_cast<ValueType>(totalNumLines) /
+          numOccurrences) + 1.0;
+    }
+  }
+
+ private:
+  //! Used to store the total number of tokens for each line.
+  std::vector<std::unordered_map<size_t, size_t>> tokensFrequences;
+  /**
+   * Used to store the number of strings which contain a token depending
+   * on the given token.
+   */
+  std::unordered_map<size_t, size_t> numContainingStrings;
+  //! Used to store the number of tokens in each line.
+  std::vector<size_t> linesSizes;
```

Comment:
Probably the encoder should clean these variables each time. Otherwise we could get different results with different encoders. Moreover these members aren't serialized.

Reply:
Would it be a decent workaround to reset these every time `InitMatrix()` is called?

Reply:
No, the encoder calls `InitMatrix()` after the first pass through the dataset. We have to introduce another function into the policy. I suggest to add something like
```c++
void Reset();
```
into each policy.

Reply:
I see, that makes sense.  Should we open an issue for it?

Reply:
I think it's easier to add the necessary changes here.

Reply:
Ok, sounds good.  I think that this PR and #2352 are the primary outstanding PRs before 3.3.0, so, let me know if I can do anything to help. :+1:  (Sorry if I keep saying that too much :smile:)

Reply:
Fixed.

Reply:
Not sure if I understand the problem, Can you explain me what the issue is 

So if we create two object - `encoder 1` and `encoder2` , they would behave differently ?

Reply:
Yes, they could behave differently if you have used one or two of them previously.